### ISSUE TITLE
DirectoryTree - Missing custom field support in typescript interface

### ISSUE THREAD

**Issue by lukasrandom** [mihneadb/node-directory-tree#103]
Hey,
just tried to add custom field to DirectoryTree object in typescript inside a `DirectoryTreeCallback`. The following error occured:

### Error:
```
error TS7053: Element implicitly has an 'any' type because expression of type '"id"' can't be used to index type 'DirectoryTree'.
  Property 'id' does not exist on type 'DirectoryTree'.

34     item['id'] = createHash('sha1').update(path).digest('base64');
```

### Example adding custom `id` field:
```
import directoryTree, {
    DirectoryTree,
    DirectoryTreeCallback,
} from 'directory-tree';
import { createHash } from 'crypto';

const path = 'C:/Users/me/Desktop/a';
const callback: DirectoryTreeCallback = (item: DirectoryTree, path: string) => {
    item['id'] = createHash('sha1').update(path).digest('base64');
};
const dirTree: DirectoryTree = directoryTree(path, {}, callback, callback);
console.log(JSON.stringify(dirTree, null, 2));
```

### Fix:
Added `[key: string]: any` in to the `DirectoryTree` interface at `node_modules\directory-tree\index.d.ts`. 
I will send you a PR so you can decide if this is a reasonable fix for it.

Best 
Lukas

**Comment by lbragile** [comment]
@lukasrandom Seems like this was merged with the fix you indicated, but even though it solves the issue you faced, I feel like that doesn't really provide meaningful typing to the custom object. 

That is, when I add properties to the `custom` object, they now all have type `any`, which doesn't help with intellisense.

Would it not be better - developer experience wise - to make the `DirectoryTree` interface (and thus `DirectoryTreeCallback` type) generic as shown below?

```ts 
declare namespace directoryTree {
  interface IObj { 
    [key: string]: unknown;
  }

  export interface DirectoryTree<C extends IObj  = IObj> {
    path: string;
    name: string;
    size: number;
    type: "directory" | "file";
    children?: DirectoryTree[];
    extension?: string;
    isSymbolicLink?: boolean;
    custom: C;
  }

  export interface DirectoryTreeOptions {
    normalizePath?: boolean;
    exclude?: RegExp | RegExp[];
    attributes?: (keyof Stats | "type" | "extension")[];
    extensions?: RegExp;
    followSymlinks?: boolean;
    depth?: number;
  }

  export type DirectoryTreeCallback<T extends IObj  = IObj> = (item: DirectoryTree<T>, path: string, stats: Stats) => void;
}
```

### Note
I haven't tested the above thoroughly, but it seems to produce desired results for me.

**Comment by mihneadb** [comment]
@lbragile sorry about the terribly late reply - I think this would complicate use more for the "average user". Thoughts of other options perhaps?

**Comment by lbragile** [comment]
@mihneadb No worries. I will have to disagree with you on:

> I think this would complicate use more for the "average user"

as my proposed typing should not cause anyone to update anything due to the same default types. That is, what I suggested only provides users the ability to "opt into" custom typing if they choose to - hence the generic interfaces.

On that note, I decided to drop this package in favor of a custom solution for my needs, so feel free to use the above if you'd like. If you do end up using it, I would appreciate a note crediting me 🙂
Or I can create a PR whichever works better for you.
